x/mongo/driver/topology/topology.go

```diff
@@ -527,7 +527,8 @@ func (t *Topology) pollSRVRecords() {
 			break
 		}
 
-		parsedHosts, err := t.dnsResolver.ParseHosts(hosts, false)
+		srvName := extractSrvServiceName(uri)
+		parsedHosts, err := t.dnsResolver.ParseHosts(hosts, srvName, false)
```

Comment:
We _could_ access `t.cfg.cs.SRVServiceName` here instead of parsing the URI again, but ideally [we should not use](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/topology_options.go#L35) the original connstring for any logic in this file. We're eventually trying to remove `topology.WithConnString` (see Jira context [here](https://jira.mongodb.org/browse/GODRIVER-1507)).

Reply:
Thank you for the context! I am not concerned with the performance overhead since polling for SRV records is infrequent.

There is less test coverage for SRV polling. My concern is that parsing the `srvServiceName` again is a more brittle. E.g. I am not sure what happens if there are duplicate options, e.g. `mongodb+srv://test1.test.build.10gen.cc/?srvServiceName=foo&srvServiceName=bar`. The URI options spec notes that [handling of duplicates is undefined](https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.rst#conflicting-tls-options). But this may risk using a different service name during polling.

Considering how other options travel from `connstring.parser#parse` to `topology.Topology`, would adding a `Topology.Option` like `WithSRVServiceName` be desirable?

Reply:
Interesting point about duplicates; the current code (`extractSrvServiceName`) would grab the first occurrence of `srvServiceName` in the URI (i.e. `foo`). The current code `connstring.parser.parse` will grab the last occurrence of `srvServiceName` in the URI (i.e. `bar`). While handling of duplicates is indeed undefined; we should at least be consistent for SRV discovery and polling. Switched `extractSrvServiceName` to a last-honored extraction.

> would a `Topology.Option` like `WithSRVServiceName` be desirable?

A good question; do we think some users are creating their own custom `Topology`s and wish to specify a custom name for only SRV polling?

Reply:
Regardless of whether users are creating `Topology`s directly, a `WithSRVServiceName` function would be very useful if it lets us remove the duplicated URI parsing code here. IIRC we should be done parsing the URI by the time we construct a `Topology` in the `Client` construction path, so I think this should work?

Reply:
That should work! I've added `Topology#WithSRVServiceName` and a new `srvServiceName` field to `topology.config`. As mentioned in an above thread, I also had to add `SetSRVServiceName` to `ClientOptions`. But now, we can remove `extractSrvServiceName` and just refer to `t.cfg.srvServiceName` here.